akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMergeSpec.scala

```diff
@@ -96,5 +96,30 @@ class FlowMergeSpec extends BaseTwoStreamsSetup {
       up1.expectSubscription().expectCancellation()
       up2.expectSubscription().expectCancellation()
     }
+
+    "not try to grab from a closed input previously enqueued" in assertAllStagesStopped {
+      // Coverage for #21138
+      val up1 = TestPublisher.probe[Int]()
+      val up2 = TestPublisher.probe[Int]()
+      val down = TestSubscriber.probe[Int]()
+
+      Source.fromPublisher(up1)
+        .merge(Source.fromPublisher(up2), eagerComplete = true)
+        .to(Sink.fromSubscriber(down))
+        .run
+
+      up1.ensureSubscription()
+      up2.ensureSubscription()
+      down.ensureSubscription()
+
+      up1.expectRequest()
+      up2.expectRequest()
+      up1.sendNext(7)
+      up2.sendNext(8)
+      up2.sendComplete()
+      down.request(1)
```

Comment:
Yeah, `8` comes in through `in2` which gets queued as out is not ready, it then cancels which closes `in2`, but the it stays in the queue and as soon as down requests an element it gets `in2` from the queue and tries to grab from it.